# Week 3: Python Basics (Part II) - Class Exercises

We first load the packages and data necessary for this exercise.

In [1]:
import pandas as pd
import numpy as np
from pydataset import data

# Load the airquality dataset
airquality = data('airquality')

initiated datasets repo at: /Users/chris/.pydataset/


---

## 1. Lottery Numbers

In order to win the top prize in a particular lottery, one must match all 6 numbers between 1 and 49 drawn by the lottery organizer.

Using the `random.choice()` function from the `numpy` library, generate **six distinct** integers between 1 and 49 (inclusive) and store the results in an object named `lottery_num`. 

Ensure the numbers do not contain any duplicates. Print the numbers in **ascending order**.

Your code should produce something like

The lottery numbers are: [14 18 28 45 46 48].

In [3]:
# Type your code here:

numbers = np.arange(1, 50)
lottery_num = np.random.choice(numbers, size=6, replace=False)
lottery_num.sort()
print(lottery_num)

[ 8 13 19 26 33 36]


---

## 2. Air Quality in New York City

In this question, we will examine daily air quality in New York City in 1973. 

The data are available through the `pydataset` package, which we installed with `requirements.txt`.

### (a) Determine the shape of the dataset and the number of unique months.
**Hint:** Have a look at `nunique` for the number of unique months.

In [5]:
# Display the first few rows of the data
airquality.head()

Ozone  Solar.R  Wind  Temp  Month  Day
1   41.0    190.0   7.4    67      5    1
2   36.0    118.0   8.0    72      5    2
3   12.0    149.0  12.6    74      5    3
4   18.0    313.0  11.5    62      5    4
5    NaN      NaN  14.3    56      5    5

In [7]:
# Find the number of rows and columns
airquality.shape

(153, 6)

In [12]:
# Find the number of unique months represented
airquality['Month'].nunique()

5

**Fill in the blanks.**

The data set has 153 (integer) rows and 6 (integer) columns. There are 5 (integer) unique months represented in the data.

#### (b) Create a `Season` column
Create a new column `Season` that takes the value of `Spring` for May, `Summer` for June, July, August, and `Autumn` for September.

Print the first 125 rows of the data.

In [13]:
# Type your code here:
conditions = [airquality['Month'].isin([5]),
              airquality['Month'].isin([6, 7, 8]),
              airquality['Month'].isin([9])]

choices = ["Spring", "Summer", "Autumn"]
airquality['Season'] = np.select(conditions, choices, default='Unknown')
airquality.head(125)

Ozone  Solar.R  Wind  Temp  Month  Day  Season
1     41.0    190.0   7.4    67      5    1  Spring
2     36.0    118.0   8.0    72      5    2  Spring
3     12.0    149.0  12.6    74      5    3  Spring
4     18.0    313.0  11.5    62      5    4  Spring
5      NaN      NaN  14.3    56      5    5  Spring
6     28.0      NaN  14.9    66      5    6  Spring
7     23.0    299.0   8.6    65      5    7  Spring
8     19.0     99.0  13.8    59      5    8  Spring
9      8.0     19.0  20.1    61      5    9  Spring
10     NaN    194.0   8.6    69      5   10  Spring
11     7.0      NaN   6.9    74      5   11  Spring
12    16.0    256.0   9.7    69      5   12  Spring
13    11.0    290.0   9.2    66      5   13  Spring
14    14.0    274.0  10.9    68      5   14  Spring
15    18.0     65.0  13.2    58      5   15  Spring
16    14.0    334.0  11.5    64      5   16  Spring
17    34.0    307.0  12.0    66      5   17  Spring
18     6.0     78.0  18.4    57      5   18  Spring
19    30.0    322.0  11.5    68      5   19  Spring
20    11.0     44.0   9.7    62      5   20  Spring
21     1.0      8.0   9.7    59      5   21  Spring
22    11.0    320.0  16.6    73      5   22  Spring
23     4.0     25.0   9.7    61      5   23  Spring
24    32.0     92.0  12.0    61      5   24  Spring
25     NaN     66.0  16.6    57      5   25  Spring
26     NaN    266.0  14.9    58      5   26  Spring
27     NaN      NaN   8.0    57      5   27  Spring
28    23.0     13.0  12.0    67      5   28  Spring
29    45.0    252.0  14.9    81      5   29  Spring
30   115.0    223.0   5.7    79      5   30  Spring
31    37.0    279.0   7.4    76      5   31  Spring
32     NaN    286.0   8.6    78      6    1  Summer
33     NaN    287.0   9.7    74      6    2  Summer
34     NaN    242.0  16.1    67      6    3  Summer
35     NaN    186.0   9.2    84      6    4  Summer
36     NaN    220.0   8.6    85      6    5  Summer
37     NaN    264.0  14.3    79      6    6  Summer
38    29.0    127.0   9.7    82      6    7  Summer
39     NaN    273.0   6.9    87      6    8  Summer
40    71.0    291.0  13.8    90      6    9  Summer
41    39.0    323.0  11.5    87      6   10  Summer
42     NaN    259.0  10.9    93      6   11  Summer
43     NaN    250.0   9.2    92      6   12  Summer
44    23.0    148.0   8.0    82      6   13  Summer
45     NaN    332.0  13.8    80      6   14  Summer
46     NaN    322.0  11.5    79      6   15  Summer
47    21.0    191.0  14.9    77      6   16  Summer
48    37.0    284.0  20.7    72      6   17  Summer
49    20.0     37.0   9.2    65      6   18  Summer
50    12.0    120.0  11.5    73      6   19  Summer
51    13.0    137.0  10.3    76      6   20  Summer
52     NaN    150.0   6.3    77      6   21  Summer
53     NaN     59.0   1.7    76      6   22  Summer
54     NaN     91.0   4.6    76      6   23  Summer
55     NaN    250.0   6.3    76      6   24  Summer
56     NaN    135.0   8.0    75      6   25  Summer
57     NaN    127.0   8.0    78      6   26  Summer
58     NaN     47.0  10.3    73      6   27  Summer
59     NaN     98.0  11.5    80      6   28  Summer
60     NaN     31.0  14.9    77      6   29  Summer
61     NaN    138.0   8.0    83      6   30  Summer
62   135.0    269.0   4.1    84      7    1  Summer
63    49.0    248.0   9.2    85      7    2  Summer
64    32.0    236.0   9.2    81      7    3  Summer
65     NaN    101.0  10.9    84      7    4  Summer
66    64.0    175.0   4.6    83      7    5  Summer
67    40.0    314.0  10.9    83      7    6  Summer
68    77.0    276.0   5.1    88      7    7  Summer
69    97.0    267.0   6.3    92      7    8  Summer
70    97.0    272.0   5.7    92      7    9  Summer
71    85.0    175.0   7.4    89      7   10  Summer
72     NaN    139.0   8.6    82      7   11  Summer
73    10.0    264.0  14.3    73      7   12  Summer
74    27.0    175.0  14.9    81      7   13  Summer
75     NaN    291.0  14.9    91      7   14  Summer
76     7.0     48.0  14.3    80      7   15  Summer
7

#### (c) Summary Statistics
Use `describe()` to display summary statistics for temperature in the summer months.

In [15]:
# Type your code here:
airquality[airquality['Season'] == 'Summer']['Temp'].describe()

count    92.000000
mean     82.358696
std       6.285482
min      65.000000
25%      78.000000
50%      82.000000
75%      86.000000
max      97.000000
Name: Temp, dtype: float64

---

## 3. Function to calculate Taxi Fare

Define a function called `get_fare(distance)` that calculates a taxi fare based on the following:
- Base fare is $5.00.
- If distance is greater than 10km, add $1.50 per km for every km **above** 10.
- Return the total fare.

Make use of your function to print 

The fare for 15km is $12.50.

In [16]:
# Type your code here:
def get_fare(distance):
    base_fare = 5.00
    if distance > 10:
        total = base_fare + (distance-10) * 1.5
    else:
        total = base_fare
    return total

print(f"The fare for 15km = {get_fare(15):.2f}")

The fare for 15km = 12.50
